evm/spec/tables/keccak-f.tex

```diff
@@ -2,3 +2,64 @@ \subsection{Keccak-f}
 \label{keccak-f}
 
 This table computes the Keccak-f[1600] permutation.
+
+\subsubsection{Keccak-f Permutation}
+To explain how this table is structured, we first need to detail how the permutation is computed. \href{https://keccak.team/keccak_specs_summary.html}{This page} gives a pseudo-code for the permutation. Our implementation differs slightly -- but remains equivalent -- for optimization and constraint degree reasons. 
+
+Let:
+\begin{itemize}
+    \item $S$ be the sponge width ($S=25$ in our case)
+    \item $NUM\_ROUNDS$ be the number of Keccak rounds (24 in our case)
+    \item $RC$ a vector of round constants of size $NUM\_ROUNDS$
+    \item $I$ be the input of the permutation, comprised of $S$ 64-bit elements
+\end{itemize}    
+
+The first step is to reshape $I$ into a $5 \times 5$ matrix. We initialize the state $A$ of the sponge with $I$: $$A[x, y] := I[x, y] \text{ }  \forall x, y \in \{0..4\}$$
+
+We store $A$ in the table, and subdivide each 64-bit element into two 32-bit limbs.
+Then, for each round $i$, we proceed as follows:
+\begin{enumerate}
+    \item First, we define $C[x] := \texttt{xor}_{i=0}^4 A[x, i]$. We store $C$ as bits in the table. This is because we need to apply a rotation on its elements' bits and carry out \texttt{ xor } operations in the next step.
+    \item Then, we store a second vector $C'$ in bits, such that: $$C'[x, z] = C[x, z] \texttt{ xor } C[x-1, z] \texttt{ xor } C[x+1, z-1]$$. 
+    \item We then need to store the updated value of $A$: $$A'[x, y] = A[x, y] \texttt{ xor } C[x, y] \texttt{ xor } C'[x, y]$$ Note that this is equivalent to the equation in the official Keccak-f description: $$A'[x, y] = A[x, y] \texttt{ xor } C[x-1, z] \texttt{ xor } C[x+1, z-1]$$.
+    \item The previous three points correspond to the $\theta$ step in Keccak-f. We can now move on the the $\rho$ and $\pi$ steps. These steps are written as: $$B[y, 2\times x + 3 \times y] := \texttt{rot}(A'[x, y], r[x, y])$$ where $\texttt{rot(a, s)}$ is the bitwise cyclic shift operation, and $r$ is the matrix of rotation offsets. We do not need to store $B$: $B$'s bits are only a permutation of $A'$'s bits. 
+    \item The $\chi$ step updates the state once again, and we store the new values: $$A''[x, y] := B[x, y] \texttt{ xor } (\texttt{not }B[x+1, y] \texttt{ and } B[x+2, y])$$ Because of the way we carry out constraints (as explained below), we do not need to store the individual bits for $A''$: we only need the 32-bit limbs.
```

Comment:
```suggestion
    \item The $\chi$ step updates the state once again, and we store the new values: $$A''[x, y] := B[x, y] \texttt{ xor } (\texttt{not }B[x+1, y] \texttt{ and } B[x+2, y])$$. Because of the way we carry out constraints (as explained below), we do not need to store the individual bits for $A''$: we only need the 32-bit limbs.
```